File: analysis/engine.py
from __future__ import annotations
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional
from ..ir.graph import Graph
from ..parsers.solidity import SolParser
from ..parsers.cairo import CaiParser
from ..detectors.base import Finding
from ..detectors.d1_sender import SenderCheckDetector
from ..detectors.d2_types import TypeMismatchDetector
from ..detectors.d3_fee import FeeDetector
from .correlator import Correlator


@dataclass
class Ctx:
    graph: Graph = field(default_factory=Graph)
    findings: List[Finding] = field(default_factory=list)
    errors: List[str] = field(default_factory=list)

    @property
    def critical(self):
        from ..detectors.base import Sev
        return any(f.sev == Sev.CRITICAL for f in self.findings)


class Engine:
    def __init__(self):
        self.detectors = [SenderCheckDetector(), TypeMismatchDetector(), FeeDetector()]
        self._hooks: List[Callable[[Finding], None]] = []

    def on_finding(self, fn): self._hooks.append(fn); return self

    def analyze(self, *, l1_path=None, l2_path=None, l1_data=None, l2_data=None) -> Ctx:
        ctx = Ctx()
        self._l1(ctx, l1_path, l1_data)
        self._l2(ctx, l2_path, l2_data)
        Correlator().run(ctx.graph)
        for d in self.detectors:
            try:
                for f in d.run(ctx.graph):
                    ctx.findings.append(f)
                    for h in self._hooks: h(f)
            except Exception as e:
                ctx.errors.append(f"[{d.did}] {e}")
        return ctx

    def _l1(self, ctx, path, data):
        try:
            p = SolParser.load(path) if path else SolParser.from_dict(data)
            for s in p.parse(): ctx.graph.add_send(s)
        except Exception as e: ctx.errors.append(f"[l1] {e}")

    def _l2(self, ctx, path, data):
        try:
            p = CaiParser.load(path) if path else CaiParser.from_dict(data)
            for h in p.parse(): ctx.graph.add_handler(h)
        except Exception as e: ctx.errors.append(f"[l2] {e}")

File: src/engine.cpp
#include "engine.hpp"
#include <fstream>
#include <sstream>
#include <filesystem>
#include <algorithm>

namespace fs = std::filesystem;

namespace crossguard {

Engine::Engine() {
    detectors.push_back(std::make_unique<SenderCheckDetector>());
    detectors.push_back(std::make_unique<TypeMismatchDetector>());
    detectors.push_back(std::make_unique<FeeDetector>());
    detectors.push_back(std::make_unique<CancellationDetector>());
    detectors.push_back(std::make_unique<ReplayRiskDetector>());
    detectors.push_back(std::make_unique<PayloadLengthDetector>());
}

AnalysisContext Engine::analyze(const std::string& l1, const std::string& l2) {
    AnalysisContext ctx;
    
    loadL1(ctx, l1);
    loadL2(ctx, l2);
    
    Correlator().run(ctx.graph);
    ctx.violations = TaintEngine().run(ctx.graph);
    
    runDetectors(ctx);
    sortFindings(ctx.findings);
    
    return ctx;
}

AnalysisContext Engine::analyzeData(const std::string& l1Data, const std::string& l2Data) {
    AnalysisContext ctx;
    
    loadL1Data(ctx, l1Data);
    loadL2Data(ctx, l2Data);
    
    Correlator().run(ctx.graph);
    ctx.violations = TaintEngine().run(ctx.graph);
    
    runDetectors(ctx);
    sortFindings(ctx.findings);
    
    return ctx;
}

void Engine::runDetectors(AnalysisContext& ctx) {
    for (auto& detector : detectors) {
        try {
            auto findings = detector->run(ctx.graph);
            for (auto& finding : findings) {
                ctx.findings.push_back(finding);
                for (auto& hook : hooks) {
                    hook(finding);
                }
            }
        } catch (const std::exception& e) {
            ctx.errors.push_back("[" + detector->id() + "] " + std::string(e.what()));
        }
    }
}

void Engine::sortFindings(std::vector<Finding>& findings) {
    std::sort(findings.begin(), findings.end(),
        [](const Finding& a, const Finding& b) {
            int aSev = static_cast<int>(a.severity);
            int bSev = static_cast<int>(b.severity);
            if (aSev != bSev) return aSev < bSev;
            return a.location.to_string() < b.location.to_string();
        });
}

void Engine::loadL1(AnalysisContext& ctx, const std::string& path) {
    try {
        std::string ext = getExtension(path);
        
        if (ext == ".sol") {
            auto parser = SoliditySourceParser::load(path);
            for (auto send : parser->parse_sends()) {
                ctx.graph.addSend(send);
            }
            for (auto canc : parser->parse_cancellations()) {
                ctx.graph.addCancellation(canc);
            }
        } else if (ext == ".json" || ext == ".ast") {
            auto parser = SolidityParser::load_json(path);
            for (auto send : parser->parse_sends()) {
                ctx.graph.addSend(send);
            }
            for (auto canc : parser->parse_cancellations()) {
                ctx.graph.addCancellation(canc);
            }
        }
    } catch (const std::exception& e) {
        ctx.errors.push_back("[L1] " + std::string(e.what()));
    }
}

void Engine::loadL1Data(AnalysisContext& ctx, const std::string& data) {
    try {
        try {
            auto parser = SolidityParser::load_json(data);
            for (auto send : parser->parse_sends()) {
                ctx.graph.addSend(send);
            }
        } catch (...) {
            auto parser = std::make_unique<SoliditySourceParser>();
            parser->source = data;
            for (auto send : parser->parse_sends()) {
                ctx.graph.addSend(send);
            }
        }
    } catch (const std::exception& e) {
        ctx.errors.push_back("[L1] " + std::string(e.what()));
    }
}

void Engine::loadL2(AnalysisContext& ctx, const std::string& path) {
    try {
        std::string ext = getExtension(path);
        
        if (ext == ".cairo") {
            auto parser = CairoSourceParser::load(path);
            for (auto handler : parser->parse_handlers()) {
                ctx.graph.addHandler(handler);
            }
        } else if (ext == ".json" || ext == ".sierra") {
            auto parser = CairoParser::load_json(path);
            for (auto handler : parser->parse_handlers()) {
                ctx.graph.addHandler(handler);
            }
        }
    } catch (const std::exception& e) {
        ctx.errors.push_back("[L2] " + std::string(e.what()));
    }
}

void Engine::loadL2Data(AnalysisContext& ctx, const std::string& data) {
    try {
        try {
            auto parser = CairoParser::load_json(data);
            for (auto handler : parser->parse_handlers()) {
                ctx.graph.addHandler(handler);
            }
        } catch (...) {
            auto parser = std::make_unique<CairoSourceParser>();
            parser->source = data;
            for (auto handler : parser->parse_handlers()) {
                ctx.graph.addHandler(handler);
            }
        }
    } catch (const std::exception& e) {
        ctx.errors.push_back("[L2] " + std::string(e.what()));
    }
}

std::string Engine::readFile(const std::string& path) {
    std::ifstream file(path);
    if (!file.is_open()) {
        throw std::runtime_error("Cannot open file: " + path);
    }
    std::stringstream buffer;
    buffer << file.rdbuf();
    return buffer.str();
}

std::string Engine::getExtension(const std::string& path) {
    size_t pos = path.find_last_of(".");
    if (pos == std::string::npos) {
        return "";
    }
    return path.substr(pos);
}

} 
